### DISCUSSION TITLE
Added: LDM notes for July 9th, 2018

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#1713] [General]
https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-07-09.md

## Agenda

1. `using var` feature
   1. Overview
   2. Tuple deconstruction grammar form
   3. `using expr;` grammar form
   4. Flow control safety

2. Pattern-based Dispose in the `using` statement

3. Relax Multiline interpolated string syntax (`$@`)

**Comment by svick** [comment]
> If C# were to allow statements on the top level, a `using System;` line could either be a using-directive if `System` is a namespace, or a using-statement if `System` is a type.

Doesn't C# already allow that, in the scripting mode?

For example, the following code currently runs fine:

```c#
using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CSharp.Scripting;

class Program
{
    static async Task Main()
    {
        string code = @"
        using System.IO;
        
        using (Stream.Null);";

        await CSharpScript.EvaluateAsync(code);
    }
}
```

**Comment by gafter** [comment]
Regarding flow-control safety:

> A new line to the spec should be added saying that, if the end of a `using` variable's lifetime is reachable, that variable must be definitely assigned at that point.

This needs to also handle constructs like this:

``` c#
{
    goto label1;
label2:
    return;
    using var x = new FileStream(...);
label1:
    goto label2;
}
```

The variable `x` must be definitely assigned at `goto label2` because `x` must be disposed by that control transfer.
